processing/src/main/java/io/druid/segment/DimensionSelector.java

```diff
@@ -42,7 +44,12 @@
    *
    * Value cardinality would be 2.
    *
-   * @return the value cardinality
+   * Cardinality may be unknown (e.g. the selector used by IncrementalIndex while reading input rows),
+   * in which case this method will return -1. If cardinality is unknown, you should assume this
+   * dimension selector has no dictionary, and avoid calling "lookupId" or "lookupName" outside of
```

Comment:
I'm usually not in favor of exceptions for "normal" control flow like this. It feels wrong to use exceptions for anything other than exceptional conditions.

I already edited the call sites I thought were the biggest potential problems (dimensionSpecs). The other call sites are in the topN engine, where it would be really bizarre for that to get called on a selector without a dictionary. But, I can edit those call sites too just so 100% of call sites are doing the proper check.

Thoughts?